analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ObjectsShouldNotBeDisposedMoreThanOnceBase.cs

```diff
@@ -18,10 +18,31 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using SonarAnalyzer.SymbolicExecution.Constraints;
+
 namespace SonarAnalyzer.SymbolicExecution.Roslyn.RuleChecks;
 
 public abstract class ObjectsShouldNotBeDisposedMoreThanOnceBase : SymbolicRuleCheck
 {
     protected const string DiagnosticId = "S3966";
-    protected const string MessageFormat = "Refactor this code to make sure '{0}' is disposed only once.";
+    protected const string MessageFormat = "Resource '{0}' has already been disposed explicitly or implicitly through a using statement. Please remove the redundant disposal.";
+
+    private static readonly string[] DisposeMethods = { "Dispose", "DisposeAsync"};
+
+    protected override ProgramState PreProcessSimple(SymbolicContext context)
+    {
+        var state = context.State;
+        if (context.Operation.Instance.AsInvocation() is { } invocation && DisposeMethods.Contains(invocation.TargetMethod.Name))
```

Comment:
There's no check for whether the object invoking `Dispose()` implements `IDisposable`.
The following code produces a FP:
```
public class NotIDisposableWithDisposeMethod
{
    public void Dispose() { }

    public void Method()
    {
        var x = new NotIDisposableWithDisposeMethod();
        x.Dispose(); // Noncompliant - FP
        x.Dispose();
    }
}
```

Reply:
To be implemented here https://github.com/SonarSource/sonar-dotnet/pull/7362

Reply:
C# 8 ref struct can implement `Dispose` method without implementing `IDisposable` interface

Reply:
It should be fine to consider any method called `Dispose`.

The rule says you should not dispose twice. So it's technically true.

Reply:
@pavel-mikula-sonarsource  I guess then I'll drop #7362